streams/shared/src/main/scala/zio/stream/StreamEffect.scala

```diff
@@ -37,68 +37,61 @@ private[stream] class StreamEffect[-R, +E, +A](val processEffect: ZManaged[R, No
 
   override def collect[B](pf: PartialFunction[A, B]): StreamEffect[R, E, B] =
     StreamEffect {
-      self.processEffect.flatMap { thunk =>
-        Managed.effectTotal { () =>
-          {
-            var b = null.asInstanceOf[B]
+      self.processEffect.map { thunk => () =>
+        {
+          var b = null.asInstanceOf[B]
 
-            while (b == null) {
-              b = pf.applyOrElse(thunk(), (_: A) => null.asInstanceOf[B])
-            }
-
-            b
+          while (b == null) {
+            b = pf.applyOrElse(thunk(), (_: A) => null.asInstanceOf[B])
           }
+
+          b
         }
       }
     }
 
   override def collectWhile[B](pred: PartialFunction[A, B]): StreamEffect[R, E, B] =
     StreamEffect {
-      self.processEffect.flatMap { thunk =>
-        Managed.effectTotal {
+      self.processEffect.map { thunk =>
+        new Function0[B] {
```

Comment:
Can you memoise the object instead of recreating it each time?

Reply:
That one, maybe, but what I really set out to fix with this PR was to see if the variable capture boxing penalty could be reduced, and for those, I don't think that memoization of the function is an option, given the mutable state.

Reply:
Ah yeah, I see. So you defining a `Function2` and then currying it wouldn't help I guess 😞

Reply:
I thought about that too haha. The thing is, as soon as you introduce a ref type to a function, regardless of whether it is an argument or result, the function is not specialized anymore.